include/pybind11/pybind11.h

```diff
@@ -1040,12 +1041,36 @@ class generic_type : public object {
 
         auto &internals = get_internals();
         auto tindex = std::type_index(*rec.type);
+        auto module_local = rec.module_local;
         tinfo->direct_conversions = &internals.direct_conversions[tindex];
-        if (rec.module_local)
+        if (module_local)
             registered_local_types_cpp()[tindex] = tinfo;
         else
             internals.registered_types_cpp[tindex] = tinfo;
-        internals.registered_types_py[(PyTypeObject *) m_ptr] = { tinfo };
+        internals.registered_types_py[type] = { tinfo };
+
+        // Clean up our internals after the Python type object gets garbage collected
+        weakref(m_ptr, cpp_function([type, tindex, module_local](handle wr) {
+            auto &internals = get_internals();
+            internals.direct_conversions.erase(tindex);
+            if (module_local)
+                registered_local_types_cpp().erase(tindex);
+            else
+                internals.registered_types_cpp.erase(tindex);
+            internals.registered_types_py.erase(type);
+            // C++20: std::erase_if(internals.inactive_override_cache,
+            //                      [type](const auto &entry) {
+            //                          return entry.first == (PyObject *) type;
+            //                      });
+            auto &cache = internals.inactive_override_cache;
+            for (auto it = cache.begin(), last = cache.end(); it != last; ) {
+                if (it->first == (PyObject *) type)
+                    it = cache.erase(it);
+                else
+                    ++it;
+            }
+            wr.dec_ref();
```

Comment:
Not sure if it'd be easily addressable, but there may be lingering references from other types here:
https://github.com/pybind/pybind11/blob/7c71dd386161703dc25860870f579cca0ff3a9a8/include/pybind11/detail/internals.h#L135-L137

e.g. you're freeing type A, but type B has a conversion related to that type

Reply:
FWIW This PR seems like a strict improvement as-is, so I'd totes be fine with a TODO / follow-up issue.

Reply:
- `implicit_casts` looks like it's only used for MI, so by construction I don't think you'd have a dangling parent base.
- `direct_conversions` looks like it's only used for NumPy dtypes
- `implicit_conversions` looks like it requires that both types be alive, so also should be fine. However, you may have unbounded memory growth.

So yeah, looks like everything is fine. Just may want to make a minor note on `implicit_conversiosn` being unbounded; however, it's unlikely that unbounded growth will matter much? (unless someone likes running the GC every 30sec for 24hrs?)

Reply:
Let me quickly check this again. I think you have a point! I checked something, but ...

Reply:
Follow-up issue in #2595, in case we don't get to this anymore before the 2.6.0 release. I'm not sure how much time I'll have over the next few days, and you're right that it is not worth it blocking the release over this! At least it's better than before :-)

Reply:
Sounds good!

Reply:
Thanks for catching these, btw; even if we don't need to fix anything, it might be nice to add a comment on why we can (currently) ignore them! :-)